### PR TITLE
[tune] Callbacks for tune runs

### LINKED ISSUE #10854: [tune] New feature: callback support for Tune
### Describe your feature request

Callbacks are a very useful addition to tuning libraries, allowing to eg. measure progress of long tuning processes. Adding the possibility to use them in Ray Tune, with, for example:

```python
tune.run(
	callbacks=[func, func2]
)
```

would be very helpful and bring it in line with other hyperparameter tuning libraries like Optuna.

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This adds a Callback API to tune. Callbacks are called from within the `TrialRunner` training loop. The base class looks like this:

```

class Callback:
    """Tune base callback that can be extended and passed to a ``TrialRunner``

    Tune callbacks are called from within the ``TrialRunner`` class. There are
    several hooks that can be used, all of which are found in the submethod
    definitions of this base class.

    The parameters passed to the ``**info`` dict vary between hooks. The
    parameters passed are described in the docstrings of the methods.
    """

    def on_step_begin(self, **info):
        """Called at the start of each tuning loop step.

        Keyword Arguments:
            iteration (int): Number of iterations of the tuning loop.
            trials (List[Trial]): List of trials.
        """
        pass

    def on_step_end(self, **info):
        """Called at the end of each tuning loop step.

        The iteration counter is increased before this hook is called.

        Keyword Arguments:
            iteration (int): Number of iterations of the tuning loop.
            trials (List[Trial]): List of trials.
        """
        pass

    def on_trial_start(self, **info):
        """Called after starting a trial instance.

        Keyword Arguments:
            iteration (int): Number of iterations of the tuning loop.
            trials (List[Trial]): List of trials.
            trial (Trial): Trial that just has been started.

        """
        pass

    def on_trial_restore(self, **info):
        """Called after restoring a trial instance.

        Keyword Arguments:
            iteration (int): Number of iterations of the tuning loop.
            trials (List[Trial]): List of trials.
            trial (Trial): Trial that just has been restored.
        """
        pass

    def on_trial_save(self, **info):
        """Called after receiving a checkpoint from a trial.

        Keyword Arguments:
            iteration (int): Number of iterations of the tuning loop.
            trials (List[Trial]): List of trials.
            trial (Trial): Trial that just saved a checkpoint.
        """
        pass

    def on_trial_result(self, **info):
        """Called after receiving a result from a trial.

        The search algorithm and scheduler are notified before this
        hook is called.

        Keyword Arguments:
            iteration (int): Number of iterations of the tuning loop.
            trials (List[Trial]): List of trials.
            trial (Trial): Trial that just sent a result.
            result (Dict): Result that the trial sent.
        """
        pass

    def on_trial_complete(self, **info):
        """Called after a trial instance completed.

        The search algorithm and scheduler are notified before this
        hook is called.

        Keyword Arguments:
            iteration (int): Number of iterations of the tuning loop.
            trials (List[Trial]): List of trials.
            trial (Trial): Trial that just has been completed.
        """
        pass

    def on_trial_fail(self, **info):
        """Called after a trial instance failed (errored).

        The search algorithm and scheduler are notified before this
        hook is called.

        Keyword Arguments:
            iteration (int): Number of iterations of the tuning loop.
            trials (List[Trial]): List of trials.
            trial (Trial): Trial that just has been completed.
        """
        pass
```

Example usage:

```
from ray import tune
from ray.tune import Callback


class MyCallback(Callback):
    def on_trial_result(self, **info):
        print(f"Got result: {info['result']['metric']}")


def train(config):
    for i in range(10):
        tune.report(metric=i)


tune.run(
    train,
    callbacks=[MyCallback()])
```

## Related issue number

Closes #10854

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by krfricke** [comment]
Main topics to discuss here:

- Do we want to support more/less hooks?
- Currently the hooks are called right after the usual processing (e.g. after talking to the scheduler/search algorithm). Is this sensible?
- Do we need to pass more parameters to some of the hooks? Which?
- Do we cover existing use cases with this implementation?

**Review by amogkam — COMMENTED** [review]
Nice, this is awesome! I think the callbacks you have already should cover everything.

File: python/ray/tune/trial_runner.py [hunk 1]
```diff
@@ -69,6 +71,119 @@ def _from_cloudpickle(self, obj):
...
+
+
+        tune.run(
+            train,
+            callbacks=[MyCallback()])
+
+    """
+
+    def on_step_begin(self, trial_runner: "TrialRunner"):
+        """Called at the start of each ``TrialRunner.step()`` invocation."""
```
Comment (amogkam) [inline python/ray/tune/trial_runner.py:106]: Regarding the docstring: As a user it's not exactly clear what TrialRunner.step() means and what exactly that entails.

File: python/ray/tune/trial_runner.py [hunk 2]
```diff
@@ -69,6 +71,119 @@ def _from_cloudpickle(self, obj):
         return cloudpickle.loads(hex_to_binary(obj["value"]))
 
 
+class Callback:
```
Comment (richardliaw) [inline python/ray/tune/trial_runner.py:74]: Should we avoid calling back into the trialrunner? the trial runner seems to be an internal component (i.e., its methods are not publicly documented).

Comment (krfricke) [reply python/ray/tune/trial_runner.py:74]: The main question then is if there are any properties that users might want to deal with. E.g. should we then pass the list of `trials`, or other pieces of the state that people might want to interact with.

Comment (richardliaw) [reply python/ray/tune/trial_runner.py:74]: Trials are good. Can we do it in a forward compat way?

Comment (krfricke) [reply python/ray/tune/trial_runner.py:74]: After our offline discussion I thought about the possibilities that we have and decided to go with a keyword argument parameter. In my opinion we retain maximum flexibility with this approach while ensuring forwards compatibility:

- Arguments can easily be added without breaking existing code
- Arguments can be deprecated and raise a warning if they are used easily
- No state in callbacks is retained, child classes can define their own constructor without having to call `super()`
- Clear documentation of keyword arguments is easier than documentation of class attributes

I updated the PR with that.

**Comment by richardliaw** [comment]
Nice! One last comment - I think it is better to pass in the known kwargs as arguments. ie:

def on_trial_save(self, iteration, trials, trial, **info):
        """Called after receiving a checkpoint from a trial.

        Keyword Arguments:
            iteration (int): Number of iterations of the tuning loop.
            trials (List[Trial]): List of trials.
            trial (Trial): Trial that just saved a checkpoint.
        """

The user will do this anyway when subclassing - and it is beneficial because you will let the ide help autocomplete.

On the other hand, you can also easily handle deprecation - you can always just check signature and warn if an old parameter is being used, while passing in the new parameter (if it is like a renaming). 

It also retains all the other benefits (ie fwd compat, good docs).

**Comment by richardliaw** [comment]
Btw can you also add a small test to make sure trial.last_result is always updated _after_ the callbacks? 

That’s the one trial property that we may want to have an explicit variant for.

**Comment by krfricke** [comment]
Re: known arguments - then what is the point of adding kwargs at all? They will be empty then. Or is it just so that the user automatically adds the `**info` kwargs to their overwritten methods so that they stay compatible in the future?

**Comment by richardliaw** [comment]
Yes, that's the idea (to add **kwargs for forward compatibility hinting).

**Comment by krfricke** [comment]
All updated - should be ready to merge after the build is passed.

**Review by richardliaw — APPROVED** [review]
